File: 1224-minimum-falling-path-sum-ii/minimum-falling-path-sum-ii.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& grid) 
    {
        int n=grid.size();
        int dp[200][200];

        for(int j=0;j<n;j++)      // populating the first column.
        dp[0][j]=grid[0][j];

        // Populate the dp array for all subsequent rows
    for (int i = 1; i < n; i++) {
        for (int j = 0; j < n; j++) {
            int minValue = INT_MAX;
            // Check all possible columns in the previous row except the same column
            for (int k = 0; k < n; k++) {
                if (k != j) {
                    minValue = min(minValue, dp[i-1][k]);
                }
            }
            dp[i][j] = grid[i][j] + minValue;
        }
    }

    // Find the minimum value in the last row
    int mini = INT_MAX;
    for (int j = 0; j < n; j++) {
        mini = min(mini, dp[n-1][j]);
    }

    return mini;
    }
};

File: 1224-minimum-falling-path-sum-ii/minimum-falling-path-sum-ii.java
class Solution {
    private int n;
    private Integer[][] dp;
    private int solve(int row, int col, int[][] grid)
    {
        if(row == n - 1)
        return grid[row][col]; 

        if(dp[row][col] != null)
        return dp[row][col];    

        int minSum = Integer.MAX_VALUE/2;
        for(int j = 0; j < n; j++)
        {
            if(j == col)
            continue;

            minSum = Math.min(minSum, solve(row + 1, j, grid));
        }

        return dp[row][col] = grid[row][col] + minSum;
    }

    public int minFallingPathSum(int[][] grid) 
    {
        n = grid.length;
        dp = new Integer[n][n];

        int res = Integer.MAX_VALUE;
        for(int col = 0; col < n; col++)
        res = Math.min(res, solve(0, col, grid));

        return res;
    }
}